### PR TITLE
[Travis] Running JavaScript tests

### LINKED ISSUE #3241: [Addressing] While editing zone, every new member renders as text input instead of select
Covered by Javascript tests. To be fixed in #3055.

### PR DESCRIPTION
**Pull request by pamil**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets | #3241
| License       | MIT

Main changes:
 - [x] Javascript based scenarios are executed since now
 - [x] Repaired a few old and rusty Behat scenarios
 - [x] Fixed #3241

Changes made by-the-way:
 - [x] MongoDB won't install if not needed
 - [x] PHP7 tests are running now (and passing, PhpSpec omitted) :tada: 
 - [x] On failure, logs and screenshots will be provided (based on [Lakion/MinkDebugExtension](https://github.com/Lakion/MinkDebugExtension))
 - [x] Added `--strict` toggle to every Behat execution (it won't pass if there are pending or missing steps)
 - [x] Introduced caching (speeds each job up for ~2-3 minutes )
 - [x] Cleaned up Travis a little bit :)

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -29,15 +35,17 @@ services:
...
     - composer install --prefer-source --no-interaction
-    - composer require doctrine/mongodb-odm="1.0.*@dev" --no-update
-    - composer require jmikola/geojson="~1.0" --no-update
-    - composer update doctrine/mongodb-odm jmikola/geojson --prefer-source --no-interaction $COMPOSER_FLAGS
+
+#    - if [ $TRAVIS_PHP_VERSION != "7.0" ]; then composer require doctrine/mongodb-odm="1.0.*@dev" --no-update; fi
+#    - if [ $TRAVIS_PHP_VERSION != "7.0" ]; then composer require jmikola/geojson="~1.0" --no-update; fi
+#    - if [ $TRAVIS_PHP_VERSION != "7.0" ]; then composer update doctrine/mongodb-odm jmikola/geojson --prefer-source --no-interaction; fi
+
+    - if [ $TEST_SUITE = "javascript" ]; then travis/prepare-to-javascript-tests; fi
```
Comment (stloyd) [inline .travis.yml]: Don't waste resources to build apache, as we changed to PHP 5.5, use built-in webserver!

```yml
before_script:
    - php -S localhost:8000 -t web &> /dev/null;

    - curl http://selenium-release.storage.googleapis.com/2.47/selenium-server-standalone-2.47.1.jar > selenium.jar
    - java -jar selenium.jar > /dev/null &
    - sleep 5
```
Later add:
```yml
# behat.yml.dist
default:
     extensions:
         Behat\MinkExtension:
            base_url: 'http://localhost:8000/app_test.php'

**Comment by stloyd** [comment]
@pamil @pjedrzejewski @michalmarcinkowski Dunno who got admin status on Travis, but __PLEASE__ don't waste resources for Sylius and cancel __ALL__ broken builds for PRs (just look how many time this PR started & blocked others)....

File: .travis.yml (hunk 1 above)
Comment (pamil) [inline .travis.yml]: I wanted to check out how it will work with Apache2, but I think it's not worth the effort. Falling back to built-in webserver.

Comment (stloyd) [inline .travis.yml]: We have done this that many times that you could avoid that ;) (just look for closed PRs with "travis" in title :stuck_out_tongue_winking_eye: )

**Comment by pamil** [comment]
@stloyd we're doing our best killing those builds, unfortunately I'm not able to do it myself, so there's some small communication delay as no one is watching builds list all the time :)

File: travis/prepare/prepare-javascript [hunk 2]
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
+# Configuring display
+sh -e /etc/init.d/xvfb start > /dev/null 2>&1
+export DISPLAY=:99
+
+# Installing google-chrome
+wget https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb > /dev/null 2>&1
+sudo dpkg -i google-chrome-stable_current_amd64.deb > /dev/null 2>&1
```
Comment (stloyd) [inline travis/prepare/prepare-javascript]: Why using chrome while firefox is already available?

Comment (pamil) [inline travis/prepare/prepare-javascript]: Primo, Firefox is much slower than Chrome (about 2-3 times slower).
Secundo, Firefox has a random issues with some pages, sometimes it scrolls down the page and there's no possibility to scroll up; when it tries to do something with element that isn't visible and accessable for that moment, it throws an exception and test fails.

Comment (pamil) [inline travis/prepare/prepare-javascript]: Btw. we may come back to container-based infrastructure, if there was a possibility to install Google Chrome on that. There's a PR for that: https://github.com/travis-ci/apt-source-whitelist/issues/35

**Comment by pjedrzejewski** [comment]
@stloyd What @pamil said, we kill them as quickly as humanly possible.

**Comment by pjedrzejewski** [comment]
@pamil you smashed it. Thank you Kamil! :+1:

**Comment by pamil** [comment]
My biggest achievement in Sylius development so far :) :tada:

File: src/Sylius/Bundle/WebBundle/Resources/views/Backend/Customer/show.html.twig [hunk 3]
```diff
@@ -19,10 +19,8 @@
         {{ buttons.manage(path('sylius_backend_customer_index'), 'sylius.customer.manage'|trans) }}
         {% if not customer.deleted %}
             {{ buttons.edit(path('sylius_backend_customer_update', {'id': customer.id})) }}
-            {% if customer.user is null %}
-                {{ buttons.delete(path('sylius_backend_customer_delete', {'id': customer.id}), null, false, false) }}
-            {% elseif customer.user.id is not sameas(app.user.id) %}
-                {{ buttons.delete(path('sylius_backend_customer_delete', {'id': customer.id}), null, false, false) }}
+            {% if customer.user is null or customer.user.id is not sameas(app.user.id) %}
```
Comment (stloyd) [inline src/Sylius/Bundle/WebBundle/Resources/views/Backend/Customer/show.html.twig]: You should use `same as`.

File: travis/prepare/prepare-mongodb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+echo "extension = mongo.so" >> ~/.phpenv/versions/$(phpenv version-name)/etc/php.ini
+
+if [ -f $SYLIUS_CACHE_DIR/composer-mongodb.lock ]; then
+    cp $SYLIUS_CACHE_DIR/composer-mongodb.lock composer.lock
+else
+    composer require doctrine/mongodb-odm="1.0.*@dev" --no-update
```
Comment (stloyd) [inline travis/prepare/prepare-mongodb]: You should use stable version.

File: features/backend/shipments.feature [hunk 5]
```diff
@@ -57,5 +57,5 @@ Feature: Shipments
          Then I should be on the shipment page with method "DHL"
 
     Scenario: Displaying the shipment state on details page
-        Given I am on the shipment page with method "DHL"
+         When I am on the shipment page with method "DHL"
```
Comment (stloyd) [inline features/backend/shipments.feature]: `Given` was proper one here IMO.

Comment (pamil) [inline features/backend/shipments.feature]: I don't think `Given / Then` workflow is correct, especially if there's `Given` included from the background.

File: travis/prepare/prepare-mongodb (hunk 4 above)
Comment (pamil) [inline travis/prepare/prepare-mongodb]: Just copied it from original Travis configuration file, feel free to contribute :)

File: src/Sylius/Bundle/WebBundle/Resources/views/Backend/Customer/show.html.twig (hunk 3 above)
Comment (pamil) [inline src/Sylius/Bundle/WebBundle/Resources/views/Backend/Customer/show.html.twig]: Same as below, just copied it, I didn't want to change more things that needed in this PR.

Comment (stloyd) [inline src/Sylius/Bundle/WebBundle/Resources/views/Backend/Customer/show.html.twig]: Problem with `sameas` it will not work with latest Twig...

Comment (pamil) [inline src/Sylius/Bundle/WebBundle/Resources/views/Backend/Customer/show.html.twig]: Do you mean Twig 2? I found that [same as is an alias for sameas](http://twig.sensiolabs.org/doc/tests/sameas.html).

Comment (stloyd) [inline src/Sylius/Bundle/WebBundle/Resources/views/Backend/Customer/show.html.twig]: https://github.com/symfony/symfony/pull/15812

Comment (pamil) [inline src/Sylius/Bundle/WebBundle/Resources/views/Backend/Customer/show.html.twig]: I'll make a PR for that :+1:

**Pull request merged**